README.md
# azurerm
Easy to use Python library for Azure Resource Manager.

The azurerm library provides wrapper functions for the Azure REST api. It doesn't include every option for every API call but it is easy to extend. The goal is to make it easy to call the API from Python using the latest API versions (in some cases before the official SDKs are available).

Note: This is not an official Microsoft library, just some REST wrappers to make it easier to call the Azure REST API. For the official Microsoft Azure library for Python please go here: <a href="https://github.com/Azure/azure-sdk-for-python">https://github.com/Azure/azure-sdk-for-python</a>.

## Latest news
For what's new in the most recent version refer to the [Changelog](https://github.com/gbowerman/azurerm/blob/master/changelog.md).

For occasional azurerm code samples and announcements see the [azurerm blog](https://msftstack.wordpress.com/?s=azurerm).

## Installation
1. pip install azurerm
2. To call these functions you need an authentication token. One way to get this is by creating a Service Principal, another is to get a bearer token using CLI. 

## Authenticating using a Service Principal
For a semi-permanent/hardcoded way to authenticate, you can create a "Service Principal" for your application (an application equivalent of a user). Once you've done this you'll have 3 pieces of information: A tenant ID, an application ID, and an application secret. You will use these to create an authentication token. For more information on how to get this information go here: [Authenticating a service principal with Azure Resource Manager](https://azure.microsoft.com/en-us/documentation/articles/resource-group-authenticate-service-principal/). See also: [Azure Resource Manager REST calls from Python](https://msftstack.wordpress.com/2016/01/05/azure-resource-manager-authentication-with-python/). Make sure you create a service principal with sufficient access rights, like "Contributor", not "Reader".

## Authenticating using CLI
When you run a CLI command, it caches an authentication token which you can use with azurerm calls. Recent versions of CLI have a command which returns an authentication token: _az account get-access-token_. Azurerm has added a new function to get the Azure authentication token from CLI's local cache: 
```
azurerm.get_access_token_from_cli()
```
This saves you from having to create a Service Princial at all. Note: This function will fail unless you have an unexpired authentication token in your local CLI cache. I.e. you have run _az login_ on the same machine recently.

Example authenticating using the Azure Portal Cloud Shell:
```
me@Azure:-$ pip install --user --upgrade azurerm
me@azure:-$ python
>>> import azurerm
>>> token = azurerm.get_access_token_from_cli()
>>> azurerm.list_subscriptions(token)
```

## azurerm examples
See below for some simple examples. A detailed set of **azurerm** programming examples can be found here: [azurerm Python library programming examples](https://github.com/gbowerman/azurerm/blob/master/examples.md). For more examples look at the [azurerm examples library](https://github.com/gbowerman/azurerm/tree/master/examples). 

For full documentation see [azurerm reference manual](https://github.com/gbowerman/azurerm/tree/master/docs).

See also the unit test suite which covers the main storage, network, compute functions - the goal is to expand it to test every function in the library: [test](https://github.com/gbowerman/azurerm/tree/master/test)

### National/isolated cloud support
To use this library with national or isolated clouds, set environment variables to override the public default endpoints.

E.g. bash shell example for China..
``` 
  export AZURE_RM_ENDPOINT='https://management.chinacloudapi.cn'
  export AZURE_AUTH_ENDPOINT='https://login.chinacloudapi.cn/'
  export AZURE_RESOURCE_ENDPOINT='https://management.core.chinacloudapi.cn/'
```

#### Example to list Azure subscriptions, create a Resource Group, list Resource Groups
```
import azurerm

# create an authentication token (use a Service Principal or call get_access_token_from_cli())
# Service principal example:
tenant_id = 'your-tenant-id'
application_id = 'your-application-id'
application_secret = 'your-application-secret'

access_token = azurerm.get_access_token(tenant_id, application_id, application_secret)

# list subscriptions
subscriptions = azurerm.list_subscriptions(access_token)
for sub in subscriptions['value']:
    print(sub['displayName'] + ': ' + sub['subscriptionId'])

# select the first subscription
subscription_id = subscriptions['value'][0]['subscriptionId']

# create a resource group
print('Enter Resource group name to create.')
rgname = input()
location = 'southeastasia'
rgreturn = azurerm.create_resource_group(access_token, subscription_id, rgname, location)
print('Create RG return code: ' + str(rgreturn.status_code)

# list resource groups
resource_groups = azurerm.list_resource_groups(access_token, subscription_id)
for rg in resource_groups['value']:
    print(rg["name"] + ', ' + rg['location'] + ', ' + rg['properties']['provisioningState'])
``` 

#### Example to create a virtual machine
See [create_vm.py](https://github.com/gbowerman/azurerm/tree/master/examples/create_vm.py).

See also an example to create a VM Scale Set [create_vmss.py](https://github.com/gbowerman/azurerm/tree/master/examples/create_vmss.py).

#### Example to create a Media Services Account
See [createmediaserviceaccountinrg.py](https://github.com/gbowerman/azurerm/tree/master/examples/createmediaserviceaccountinrg.py)

## Functions currently supported
A basic set of infrastructure create, list, query functions are implemented. If you want to add something please send me a PR (don't forget to update this readme too).

See the [Function reference](https://github.com/gbowerman/azurerm/tree/master/docs) for full documentation.

changelog.md
# azurerm - change log
### v0.10.0 (5/1/19):
- New Compute API version 2019-03-01
- Adding skip_shutdown boolean option to stop_vm() - True will power-off VM in 10s or less

### v0.9.18 (4/18/19): Adding list nsg/asg functions - thanks @mfunger
- list_nsgs(access_token, subscription_id, resource_group)
- list_nsgs_all(access_token, subscription_id)
- list_asgs(access_token, subscription_id, resource_group)
- list_asgs_all(access_token, subscription_id)

### v0.9.15 (12/13/18):
- Fix to graph functions (incorrect GRAPH_RESOURCE_HOST value)

### v0.9.14 (12/13/18):
- Fix to graph functions (required import os library)

### v0.9.13 (12/13/18):
- Added Microsoft Graph functions: 
  get_graph_token_from_msi(), get_object_id_from_graph()
- A user's object ID is required when creating an Azure Key Vault for example.
- get_graph_token_from_msi() assumes you're running in an Azure Cloud Shell or VM.

### v0.9.12 (12/11/18):
- Added list_tenants() to list tenants user has access to. Added test and updated 
  subscriptions_test.py. Updated BASE_API version.
- Added key vault functions: create_keyvault(), delete_keyvault(), get_keyvault(),
  list_keyvaults(), list_keyvaults_sub(). Added keyvault unit tests.
  
### v0.9.11 (11/08/18):
- Update get_access_token_from_cli() to work in Azure cloud shell by getting token from MSI
  endpoint

### v0.9.10 (11/07/18):
- BREAKING CHANGES
   get_storage_usage() now takes a location parameter
   list_vm_instance_view() is removed as the instance view is now provided in the list_vm call.
- COMP_API version lowered to 2018-06-01 as 10-01 doesn't support listing VMSS NICs
- Updated QA tests to introduce some delays (creating NSG, subnet, storage acc)

### v0.9.9 (11/07/18):
- Updated STORAGE_API to 2018-07-01

### v0.9.7 (11/06/18):
- Added DEPLOYMENT_API version 2018-05-01
- Updated COMP_API to 2018-10-01 and NETWORK_API to 2018-08-01

### v0.9.6 (12/11/17):
- Incremented Resource groups API version to 2017-05-10
- Added function: export_template(access_token, subscription_id, rgname) + unit test
- Added function: get_resource_group_resources(access_token, subscription_id, rgname) + unit test

### v0.9.5 (11/8/17):
- Incremented Compute API version to 2017-12-01

### v0.9.4 (11/6/17):
- Adding media services examples (msleal)
- Fixing get_subscription_from_cli() bug (msleal)

### v0.9.3 (10/16/17):
- BREAKING CHANGE: 
  create_vmss() lb pool parameters are now optional.
  New usage: 
  ```
  create_vmss(access_token, subscription_id, resource_group, vmss_name, vm_size, capacity,
              publisher, offer, sku, version, subnet_id, location, be_pool_id=None,
              lb_pool_id=None, storage_type='Standard_LRS', username='azure', password=None,
              public_key=None, overprovision=True, upgrade_policy='Manual',
              public_ip_per_vm=False)
  ```
  location arg is now before LB args, which can be omitted if you're not using a load balancer.
  
### v0.9.2 (9/19/2017): 
- Media services fixes. Thanks @msleal.
- create_asset_delivery_policy() and get_key_delivery_url()

### v0.9.1 (9/18/2017): 
- Add put_vmss() and put_vmss_vm() functions for easy updating of scale set properties.
- Note: put_vmss_vm() is currently preview and won't work outside of Canary environments.

### v0.9.0 (9/16/2017): 
- Integration with msleal's Azure Media Services API wrapper library.
- Part 1: integration of core APIs. Example library integration planned.

### v0.8.28 (8/29/17):
- BREAKING CHANGE: 
  Improved scale_vmss() to no longer require VM size or sku - now it only requires resource group,
  VMSS name and capacity. New definiton:
  def scale_vmss(access_token, subscription_id, resource_group, vmss_name, capacity)

### v0.8.27 (8/28/17):
- Fixed list_vmss_nics() (incorrect API version)

### v0.8.26 (8/27/17):
- Fixed list_insights_components()

### v0.8.25 (8/22/17):
- Comments formatting for README.py only.

### v0.8.24 (8/22/17):
- Added cosmosdb API - thanks @iainfoulds.

### v0.8.23 (8/11/17):
- Auto-generated documentation now in place. See [azurerm Technical Reference](https://github.com/gbowerman/azurerm/tree/master/docs).
- Documentation generated by [py2md](https://github.com/gbowerman/py2md).

### v0.8.22 (8/10/17):
- Breaking change - container instance group functions renamed to xxx_container_instance_group_yyy
   e.g. create_container_instance_group(), list_container_instance_groups_sub() etc.
- Complete docstring comments work. All functions have complete docstring descriptions.

### v0.8.20 (8/04/17):
- Comments improvements, using docstring, with a view to hosting docs on readthedocs soon.
- Made get_access_key_from_cli() check all keys in the CLI cache.

### v0.8.18 (8/04/17):
- BREAKING CHANGE:
- Changed how container instance groups are handled, in order to manage multiple container instances in a group
- Added helper function: create_container_definition(container_name, image, port=80, cpu=1.0, memgb=1.5, environment=None) 
- Create a list of these definitions which is then passed to create_container_group()
- changed other container instance functions to work at a group level
- create_container_group(access_token, subscription_id, resource_group, container_group_name,
                              container_list, location, ostype='Linux', port=80, iptype='public')
- delete_container_group(access_token, subscription_id, resource_group, container_group_name)
- get_container_group(access_token, subscription_id, resource_group, container_group_name)
- get_container_logs(access_token, subscription_id, resource_group, container_group_name, container_name=None)
- list_container_groups(access_token, subscription_id, resource_group)
- list_container_groups_sub(access_token, subscription_id)

### v0.8.17 (8/04/17):
- ISO time format fix for get_access_token_from_cli()

### v0.8.16 (8/04/17):
- Added error checking to get_access_token_from_cli() to check accessToken hasn't expired

### v0.8.15 (8/04/17):
- BREAKING CHANGE:

azurerm/settings.py
'''settings.py - place to store constants for azurerm'''
import os

#AMS Endpoints...
ams_auth_endpoint = 'https://wamsprodglobal001acs.accesscontrol.windows.net/v2/OAuth2-13'
ams_rest_endpoint = 'https://media.windows.net/API'

# public defaults for authentication and resource endpoints
AZURE_RM_ENDPOINT = 'https://management.azure.com'
AZURE_AUTH_ENDPOINT = 'https://login.microsoftonline.com/'
AZURE_RESOURCE_ENDPOINT = 'https://management.core.windows.net/'
GRAPH_RESOURCE_HOST = 'graph.microsoft.com'

ACS_API = '2017-01-31'
BASE_API = '2016-06-01'
COMP_API = '2019-03-01'
CONTAINER_API = '2017-08-01-preview'
COSMOSDB_API = '2015-04-08'
DEPLOYMENTS_API = '2018-05-01'
INSIGHTS_API = '2015-04-01'
INSIGHTS_COMPONENTS_API = '2015-05-01'
INSIGHTS_METRICS_API = '2016-03-01'
INSIGHTS_PREVIEW_API = '2016-06-01'
KEYVAULT_API = '2018-02-14'
MEDIA_API = '2015-10-01'
NETWORK_API = '2018-08-01'
RESOURCE_API = '2017-05-10'
STORAGE_API = '2018-07-01'

# AMS Headers
json_only_acceptformat = "application/json"
json_acceptformat = "application/json;odata=verbose"
xml_acceptformat = "application/atom+xml"
batch_acceptformat = "multipart/mixed" 
xmsversion = "2.13"
dsversion_min = "3.0;NetFx"
dsversion_max = "3.0;NetFx"
charset = "UTF-8"


def get_rm_endpoint():
    '''Set Azure Resource Manager endpoint by environment variable, else return default value.

    These functions facilitate the use of national and isolated clouds by allowing endpoints to
    be set dynamically. The default settings, if no environment varibles are used,
    are for public cloud.
    '''
    rm_endpoint = os.environ.get('AZURE_RM_ENDPOINT')
    if rm_endpoint is None:
        return AZURE_RM_ENDPOINT
    else:
        return rm_endpoint


def get_auth_endpoint():
    '''Set Azure auth endpoint by environment variable, else return default value.
    '''
    auth_endpoint = os.environ.get('AZURE_AUTH_ENDPOINT')
    if auth_endpoint is None:
        return AZURE_AUTH_ENDPOINT
    else:
        return auth_endpoint


def get_resource_endpoint():
    '''Set Azure reosurce endpoint by environment variable, else return default value.
    '''
    resource_endpoint = os.environ.get('AZURE_RESOURCE_ENDPOINT')
    if resource_endpoint is None:
        return AZURE_RESOURCE_ENDPOINT
    else:
        return resource_endpoint

